File: backend/app/wholesale/services/currency.py
"""Foreign-currency money on wholesale order lines, voucher lines, and receiving costs.

Every stored money field in the wholesale tables (``selling_price``, ``buying_price``,
``amount``) stays a Kyat value — every downstream reader (order/voucher totals, Finance
balances, inventory valuation, cost/profit reports, monitoring) keeps summing that one
column and needs no changes, mixed-currency records included. A line entered in a
foreign currency additionally keeps the original amount and the exchange rate it was
converted at, as an immutable snapshot on that row: the rate is fixed the moment the
line is saved, never re-looked-up from Settings' "today's rate" later, so editing an old
foreign line does not silently reprice it and changing today's rate affects only new
lines created after the change.

This module is the one place that validates a currency/original/rate combination and
computes the Kyat amount actually stored — callers (the three ``_line``/cost builders)
must not compute the conversion themselves, so the server, not the browser, is
authoritative for it.
"""

from decimal import ROUND_HALF_UP, Decimal

from fastapi import HTTPException, status

DEFAULT_CURRENCY = "MMK"

# The currencies this business actually deals in. Add a code here (and to Settings'
# today_exchange_rates default in app/models/app_settings.py) when a new supplier
# currency comes up — anything else is rejected rather than silently accepted with a
# rate of who-knows-what.
SUPPORTED_CURRENCIES = {"MMK", "THB"}


def _to_decimal(value) -> Decimal:
    return value if isinstance(value, Decimal) else Decimal(str(value))


def kyat_amount(original_amount, exchange_rate) -> Decimal:
    """``original_amount`` x ``exchange_rate``, rounded to the nearest Kyat the same way
    every other money field in this app is — two decimal places, half rounds up."""
    return (_to_decimal(original_amount) * _to_decimal(exchange_rate)).quantize(
        Decimal("0.01"), rounding=ROUND_HALF_UP
    )


def resolve_money(
    currency_code: str | None,
    mmk_amount,
    original_amount,
    exchange_rate,
) -> tuple[str, Decimal, Decimal | None, Decimal | None]:
    """Validate one money field's currency/original/rate combination and return what to
    actually store: ``(currency_code, kyat_amount, original_amount, exchange_rate)``.

    MMK (the default): ``original_amount``/``exchange_rate`` must be absent, and
    ``mmk_amount`` — exactly what was typed — is stored as-is. Any other currency:
    both ``original_amount`` and a positive ``exchange_rate`` are required, and the
    Kyat amount actually stored is computed from them here rather than from whatever
    ``mmk_amount`` the client sent, since a foreign line's Kyat figure is a derived
    snapshot, not something typed directly.
    """
    code = (currency_code or DEFAULT_CURRENCY).strip().upper()
    if code not in SUPPORTED_CURRENCIES:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, f'Unsupported currency "{code}"')

    if code == DEFAULT_CURRENCY:
        if original_amount is not None or exchange_rate is not None:
            raise HTTPException(
                status.HTTP_422_UNPROCESSABLE_CONTENT,
                "An MMK amount cannot also carry an original amount or exchange rate",
            )
        if mmk_amount is None:
            raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "An MMK amount is required")
        return code, _to_decimal(mmk_amount), None, None

    if original_amount is None or exchange_rate is None:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_CONTENT,
            f"A {code} amount needs both an original amount and an exchange rate",
        )
    rate = _to_decimal(exchange_rate)
    if rate <= 0:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "Exchange rate must be greater than zero")
    original = _to_decimal(original_amount)
    if original < 0:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_CONTENT, "Original amount cannot be negative")
    return code, kyat_amount(original, rate), original, rate

File: frontend/renderer/src/components/features/wholesale/shared/currency.ts
// Foreign-currency money on customer-order lines, supplier-voucher lines, and
// receiving costs — mirrors backend/app/services/wholesale/currency.py, the one place
// that actually validates a currency/original/rate combination and computes the Kyat
// amount. Everything here is client-side preview only: the server recomputes and is
// authoritative, so a rounding difference between this file and the backend is
// cosmetic, never a source of truth mismatch.

export const DEFAULT_CURRENCY = "MMK";

/** The currencies this business actually deals in, alongside MMK. Keep this list in
 *  step with backend/app/services/wholesale/currency.py::SUPPORTED_CURRENCIES and
 *  Settings' exchange-rate section — adding one here without the backend change means
 *  a line can be typed but never saved. */
export const FOREIGN_CURRENCIES = ["THB"] as const;
export type ForeignCurrency = (typeof FOREIGN_CURRENCIES)[number];
export type CurrencyCode = "MMK" | ForeignCurrency;
export const CURRENCY_CODES: CurrencyCode[] = [DEFAULT_CURRENCY, ...FOREIGN_CURRENCIES];

export function isForeignCurrency(code: string): code is ForeignCurrency {
  return (FOREIGN_CURRENCIES as readonly string[]).includes(code);
}

/** A money field's full state, as carried on an order line / voucher line / receiving
 *  cost. `amount` is always the Kyat figure, the one every total/report reads; the rest
 *  is only present for a non-MMK currency, and is a fixed snapshot once saved. */
export interface MoneySnapshot {
  currency_code: string;
  original_amount: number | null;
  exchange_rate: number | null;
}

/** original_amount x exchange_rate, rounded to the nearest Kyat the same way the
 *  server rounds it — two decimal places, half rounds up. A client-side preview only;
 *  the server recomputes this from the same two numbers and is authoritative. */
export function previewKyatAmount(originalAmount: number, exchangeRate: number): number {
  if (!Number.isFinite(originalAmount) || !Number.isFinite(exchangeRate)) return 0;
  return Math.round(originalAmount * exchangeRate * 100) / 100;
}

const RATE_FORMAT = new Intl.NumberFormat("en-US", { maximumFractionDigits: 6 });

export function formatRate(rate: number): string {
  return RATE_FORMAT.format(rate);
}

export function formatOriginalAmount(currencyCode: string, amount: number): string {
  return `${currencyCode} ${new Intl.NumberFormat("en-US", { maximumFractionDigits: 2 }).format(amount)}`;
}
